[#15] 마이페이지 쿠폰 조회 & 사용 가능한 쿠폰 조회 & 쿠폰 사용

codesejin opened the pull request:
- 관련 Issue : https://github.com/f-lab-edu/off-coupon/issues/15

## 테이블 증설

쿠폰 사용 로직에 주문 비즈니스 로직이 포함되어야 할것 같아서 테이블 3개를 추가했습니다.

<img width="766" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/b9eaa4a8-b5cc-4f3f-9f65-7d7c391effcb">

# 궁금한 점

`쿠폰을 사용한 주문 API`에서 `DB 조회`가 많이 일어나고 있는것 같은데, 관련 데이터는 한번에 받아서 DTO에 담아야 할지, 
아니면 간단한 SELECT쿼리는 DB 부하에 영향을 크게 주지 않기 때문에 현재로서도 충분한지 궁금합니다.

codesejin committed:
docs : README.md 수정

codesejin committed:
docs : Update README.md

f-lab-bot committed:
자동 PR 리뷰를 위한 workflow 파일 추가 by f-lab

codesejin committed:
chore : "JPA"에서 "Mybatis"로 변경

Subquery, Join 등 코드 리뷰를 더 세심하게 받기 위함

codesejin committed:
Merge pull request from master

codesejin committed:
docs : Read Me ERD 등 추가

codesejin committed:
feat : 마이페이지 갖고 있는 쿠폰 조회 & 주문 시 사용 가능한 쿠폰 조회

codesejin committed:
refactor : 쿠폰 사용 가능 검증 여부 리팩토링

codesejin committed:
feat : 쿠폰 사용 비즈니스 로직 구현

codesejin committed:
refactor : list empty체크

codesejin committed:
refactor : 쿠폰 사용 검증 관련 익셉션 처리

codesejin committed:
refactor : 사용 가능한 쿠폰 목록 조회 리팩토링

codesejin committed:
refactor : java doc 추가

codesejin committed:
refactor : 불필요한 코드 제거

codesejin committed:
refactor : QueryTest @Value사용 방법 적용 못함

codesejin committed:
refactor : 스키마 수정

codesejin committed:
refactor : 스키마 수정

codesejin committed:
refactor : 스키마 수정

codesejin committed:
chore : properties Value주입 안되서 테스트하느라 직접 명시한거 수정

codesejin committed:
refactor : 테스트하느라 @Scheduled 주석 처리한거 원복

codesejin committed:
refactor : 메소드명, 변수명 명확하게 변경

f-lab-maverick commented on src/main/java/com/flab/offcoupon/controller/MyPageController.java:
```diff
@@ -0,0 +1,24 @@
+    private final MyPageService myPageService;
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping("/coupons")
+    public ResponseEntity<ResponseDTO<List<AllCouponsByMemberIdResponse>>> getAllCoupons(@RequestParam final long memberId) {
```
자기자신의 쿠폰 리스트를 확인하는데, 파라미터로 memberId를 받을 필요가 있을까요?
이 유저가 다른 유저의 id를 파라미터로 입력해서 데이터를 조회한다면 어떻게 될까요?

별도의 권한체크가 없으니 parameter manipulation 공격이 성립하겠네요. 이 키워드를 한번 찾아보시고, 취약점을 방어해주세요.

f-lab-maverick replied:
이 부분은 이번 PR에서 수정하지 않고 미루기로했으니, 우선 해당 URL이 오픈되지 않도록 핸들러매핑이 연결되지 않게  `@GetMapping("/coupons")`을 지워주세요~~

codesejin replied:
네엡 다음번 PR에서 수정하도록하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/controller/OrderController.java:
```diff
@@ -0,0 +1,52 @@
+     * @param productId 상품 ID
+     * @return 사용 가능한 쿠폰 목록
+     */
+    @ResponseStatus(HttpStatus.OK)
+    @GetMapping("/available-coupons")
+    public ResponseEntity<ResponseDTO<List<AvailableCouponsByMemberIdResponse>>> getAvailableCoupons(@RequestParam final long memberId,
+                                                                                                     @RequestParam final long productId) {
```
이 부분도 마찬가지로, 자기자신의 쿠폰 목록을 조회하는 것인데 `memberId`를 파라미터로 받을 필요가 있을까요?

f-lab-maverick commented on src/main/java/com/flab/offcoupon/controller/OrderController.java:
```diff
@@ -0,0 +1,52 @@
+     * @param productId 상품 ID
+     * @return 사용 가능한 쿠폰 목록
+     */
+    @ResponseStatus(HttpStatus.OK)
```
`ResponseStatus`는 사용에 주의해야 합니다. Return type이 `ResponseEntity`이며, 어노테이션 `ResponseStatus`를 동시에 쓰는 상황에서 두가지 반환값이 서로 다르다면 어떤 status가 반환될까요?

codesejin replied:
Controller에서 ResponseEntity와 @ResponseStatus를 동시에 사용할때에는 ResponseEntity에서 명시된 상태코드가 우선됩니다. 즉, ResponseEntity에서 명시된 상태코드가 HTTP 응답으로 반환됩니다.


하지만 응답이 2번 발생할 수 있기 때문에 동시에 2개를 사용하는 것은 권장되지 않고, 아래와 같이 (제가 한 실수,,) 응답코드를 Created와 Ok를 둘 다 사용한다면 다른 개발자가 코드를 읽을때 혼동을 줄 수도 있습니다. 따라서 휴먼 에러를 방지하기 위해 왠만하면 `Controller에서  @ResponseStatus를 사용하지 않도록 수정하겠습니다!`
<img width="1153" alt="image" src="https://github.com/f-lab-edu/off-coupon/assets/101460733/df138515-d18a-4763-b752-bfd6af464651">

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/entity/Product.java:
```diff
@@ -0,0 +1,25 @@
+    private final String title;
+    private final String description;
+    private final long originalPrice; // 원래 가격
+    private final Long salePrice; // null일 경우 전체 할인이 적용되지 않은 것으로 간주
```
null로 비즈니스 로직을 표현하는 것은 NPE의 가능성을 높입니다.
이런 경우 salePrice를 0으로 기본값을 표기해준다면 어떨까요?

codesejin replied:
정밀도 향상을 위해 BigDecimal타입으로 변경했고, default 0으로 설정했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/coupon_use/OrderService.java:
```diff
@@ -0,0 +1,132 @@
+    public ResponseDTO<List<AvailableCouponsByMemberIdResponse>> getAvailableCoupons(final long memberId, final long productId, final LocalDateTime now) {
+        List<AvailableCouponsByMemberIdVo> availableCoupons =
+                couponIssueRepository.getAvailableCoupons(new MemberIdProductIdNowVo(memberId, productId, now));
+        List<AvailableCouponsByMemberIdResponse> responseList = new ArrayList<>();
+        filterAvailableCouponsWithMinPrice(productId, availableCoupons, responseList);
```
Collection을 생성한 후 메서드 내부에서 컬랙션의 상태를 변경시키고 있네요.
이렇게 내부에서 상태를 변경시키는 메서드는 코드를 읽을 때 해당 변수에 대해서 계속해서 트래킹하게 만들기 때문에 피로감을 느끼게 만들고, 메서드의 결과값을 예측할 수 없게 만듭니다.

그렇기 때문에 상태를 변경시키는 것이 아닌, 행위 그 자체를 메서드화한다면 좀 더 깔끔하고 읽기 쉬운 직관적인 코드로 만들 수 있습니다.

```suggestion
        List<AvailableCouponsByMemberIdResponse> responseList = availableCoupons.stream
        .filter(coupon -> isOverThenMinPrice(productId))
        ...
```

codesejin replied:
말씀해주신 filter를 사용해보려고 했는데, 할인 가격이 minOrderPrice를 초과하지 않는 범위까지만 누적하는거라 filter 사용이 어려울것 같습니답...

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/coupon_use/OrderService.java:
```diff
@@ -0,0 +1,132 @@
+     * @param availableCoupons 사용 가능한 쿠폰 목록
+     * @param responseList     결과 리스트
+     */
+    private void filterAvailableCouponsWithMinPrice(long productId, List<AvailableCouponsByMemberIdVo> availableCoupons, List<AvailableCouponsByMemberIdResponse> responseList) {
+        // min_price(상품의 최소 주문 금액)과 비교하여 조건에 맞는 쿠폰만 결과 리스트에 추가합니다.
+        if (!availableCoupons.isEmpty()) {
+            long accumulatedDiscountPrice = 0;
+            for (AvailableCouponsByMemberIdVo availableCoupon : availableCoupons) {
+                // 최소 주문 금액까지 할인 가능한 금액 누적
+                accumulatedDiscountPrice += availableCoupon.discountedPrice();
+                // 할인 금액이 min_price보다 작거나 같으면 결과 리스트에 추가
+                if (accumulatedDiscountPrice <= productRepository.getProductMinOrderPriceById(productId)) {
+                    AvailableCouponsByMemberIdResponse response = new AvailableCouponsByMemberIdResponse(availableCoupon);
+                    responseList.add(response);
+                }
+            }
+        }
+    }
+
```
메서드 내부 코드의 depth가 상당히 깊습니다.
코드블록이 깊어지면 깊어질수록 이 코드가 어떤 조건일 때 수행되는지 판단하기 어렵습니다. 심해지면 (재밌는 짤이지만) 아래와 같은 코드가 나오기도 합니다. 아래와 같은 코드일 때 어떤 조건에서 이 코드가 수행되는지 알 수 있을까요?
![image](https://github.com/f-lab-edu/off-coupon/assets/131429812/6b1c32df-0a41-4ab7-97d8-45e23fd1da67)

그렇기 때문에 코드를 작성하실 때에는 코드가 중첩된상태가 되지 않도록 주의하며, 여러 기법을 사용해야 합니다.
그 기법중 하나는 early return pattern 입니다. 아래와 같이 하나의 코드 중첩을 없앨 수 있어요. 어떻게 하면 코드 깊이를 줄일 수 있을지 고민하며 한번 리팩토링해보세요~

```suggestion
    private void filterAvailableCouponsWithMinPrice(long productId, List<AvailableCouponsByMemberIdVo> availableCoupons, List<AvailableCouponsByMemberIdResponse> responseList) {
        // min_price(상품의 최소 주문 금액)과 비교하여 조건에 맞는 쿠폰만 결과 리스트에 추가합니다.
        if (availableCoupons.isEmpty()) {
          return List.of();
        }
           ...
        }
    

```

codesejin replied:
인덴트를 줄이기 위해 람다표현식과 AtomicReference를 사용하는걸로 변경했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/coupon_use/OrderService.java:
```diff
@@ -0,0 +1,132 @@
+            orderCouponRepository.save(createOrderCoupon(orderDetailId, couponInfo));
+        }
+
+        // 5. 쿠폰 사용 처리
+        couponIssueRepository.updateCouponStatus(request.getCouponIssueId());
```
메서드 명인 `updateCouponStatus`는 쿠폰을 사용했다고 알아차리기 어려운 네이밍이네요~

codesejin replied:
아직 메소드 명이 뭐가 적절한지 판단하는게 미숙한것 같습니다! `useCoupon` 으로 변경하겠습니다 감사합니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/DiscountUtils.java:
```diff
@@ -0,0 +1,107 @@
+ */
+@UtilityClass
+public class DiscountUtils {
+    /**
+     * 할인 정보를 반환하는 메소드입니다.<br>
+     * <p>
+     * String 더하기 연산은 성능이 좋지 않을 수 있으므로 StringBuilder를 사용하여 최적화하였습니다.
+     *
+     * @param discountType  할인 종류 (DiscountType enum 값)
+     * @param discountRate  할인율 (percent 할인의 경우)
+     * @param discountPrice 할인액 (amount 할인의 경우)
+     * @return 할인 정보를 나타내는 문자열
+     */
+    public String getDiscount(DiscountType discountType, Long discountRate, Long discountPrice) {
+        StringBuilder sb = new StringBuilder();
+        if (discountType == DiscountType.AMOUNT) {
+            return sb.append(discountPrice).append("원 할인").toString();
+        } else {
+            return sb.append(discountRate).append("% 할인").toString();
+        }
+    }
```
반복문이 아닌 곳에서의 덧셈 연산은 최적화가 되기 때문에 별도로 `StringBuilder`를 사용할 필요는 없습니다. 어떤 방식으로 최적화가 되기 때문일까요?

codesejin replied:
앗 잠시 착각 했던것 같습니다! 주석에 적힌 최적화 내용은 변경하도록 하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/DiscountUtils.java:
```diff
@@ -0,0 +1,107 @@
+        }
+    }
+
+    /**
+     * 상품의 개별 가격을 반환하는 메소드입니다.<br>
+     * 만약 세일 가격이 없다면 원래 가격을 반환합니다.
+     *
+     * @param product 상품 정보
+     * @return 상품의 개별 가격
+     */
+    public long getProductPricePerUnit(Product product) {
+        return product.getSalePrice() == null ? product.getOriginalPrice() : product.getSalePrice();
+    }
+
```
이런 도메인에 종속된 비즈니스로직은 유틸리티가 아닌 도메인 클래스 내부에 메서드를 위치시켜야 응집도를 높일 수 있습니다. 누군가 도메인을 살펴볼 때 유틸클래스를 보지 않는 이상 이런 분기를 찾기 어려워지기 때문입니다.

이런식으로 운용할 수 있지 않을까요?
```
class Product {
  ...
  public long getPrice() {
    if (product.getSalePrice() == null) {
      return originalPrice;
    }
    return salePrice;
  }
}
```

codesejin replied:
메소드를 분리한다는 생각만 하고, 도메인 로직인걸 잊고 있었습니다! 말씀해주신대로 반영하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/DiscountUtils.java:
```diff
@@ -0,0 +1,107 @@
+        return product.getSalePrice() == null ? product.getOriginalPrice() : product.getSalePrice();
+    }
+
+    /**
+     * 상품의 총 가격을 계산하여 반환하는 메소드입니다.
+     *
+     * @param product  상품 정보
+     * @param quantity 상품 수량
+     * @return 상품의 총 가격
+     */
+    public long calculateTotalPrice(Product product, long quantity) {
+        return getProductPricePerUnit(product) * quantity;
+    }
```
이 로직또한 비즈니스 로직으로, 유틸리티가 아니기 때문에 Service / Domain Layer에 존재해야하는 로직입니다.

codesejin replied:
계산하는 로직이 코드 중복될 가능성이 있다고 생각해서 별도의 클래스에 분리하려고 시도했는데, 제가 `유틸리티` 라는 개념에 대해 미숙했던것 같습니다 수정하겠습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/util/DiscountUtils.java:
```diff
@@ -0,0 +1,107 @@
+        long totalDiscountPrice = 0;
+        for (Coupon coupon : couponList) {
+            if (coupon.getDiscountType() == DiscountType.PERCENT) {
+                totalDiscountPrice += getProductPricePerUnit(product) * ((double) coupon.getDiscountRate() / 100);
```
값을 double type으로 직접 계산하고 있네요. 숫자를 이렇게 원시 타입으로 계산하면 어떤 일이 발생할까요? 원하는 값이 정확하게 나올까요?

codesejin replied:
double형태로 계산하면 overflow 와 부동 소수점 이슈가 발생합니다. BigDecimal로 변경해서 수정하겠습니다!

f-lab-maverick commented on src/main/resources/mapper/CouponIssueMapper.xml:
```diff
@@ -49,4 +52,76 @@
+               e.description,
+               ci.id as couponIssueId,
+               ci.coupon_status,
+               ROUND(CASE
+                         WHEN p.sale_price IS NULL THEN
+                                 p.original_price -
+                                 IF(c.discount_type = 'PERCENT', p.original_price * (c.discount_rate / 100),
+                                    c.discount_price)
+                         ELSE
+                                 p.sale_price - IF(c.discount_type = 'PERCENT', p.sale_price * (c.discount_rate / 100),
+                                                   c.discount_price)
+                         END, 0)
+                    AS discounted_price
```
query에 로직이 있네요~~
정말 불가피한 경우가 아니라면 아래 케이스는 모두 지양해야 합니다.
- query에서 분기를 태우는 case-when-then
- query에서 값을 계산하는 경우
- query에서 비즈니스 로직이 있는 경우

왜냐하면, query는 디버깅하기가 어려우며 수정사항이 발생하였을 경우 유동적으로 변경할 수 없고, 테스트도 짜기 어렵기 때문입니다. 데이터 조회는 심플하게 조회하고, 모든 가공은 어플리케이션에서 진행해주세요~

codesejin replied:
와! 제가 항상 고민했지만 확신이 없던 내용인데, 이렇게 원인과 케이스별로 정리해주셔서 감사합니다! 애플리케이션 내에서 가공하도록 수정하겠습니다!

f-lab-maverick commented on src/main/resources/mapper/CouponIssueMapper.xml:
```diff
@@ -49,4 +52,76 @@
+                 JOIN event e ON e.id = c.event_id
+                 JOIN product p ON p.category = e.category
+        WHERE ci.member_id = #{memberId}
+          AND c.validate_start_date <= #{currentDateTime}
+          AND c.validate_end_date >= #{currentDateTime}
```
`BETWEEN` 문을 쓸 수 있엤네요.

codesejin replied:
네엡 변경했습니다 감사합니다~

f-lab-maverick commented on src/main/resources/mapper/CouponMapper.xml:
```diff
@@ -50,13 +72,29 @@
+
+    <select id="validateNowIsBetweenPeriod" parameterType="java.util.HashMap" resultType="ValidateNowIsBetweenPeriodVo">
+    <![CDATA[
+        select id        as couponId,
+               if(validate_start_date <= #{currentDateTime} and validate_end_date >= #{currentDateTime}, true,
+                  false) as isBetweenValidatePeriod,
```
이 부분도 쿼리에 로직이 들어간 케이스로 보이네요~

codesejin replied:
애플리케이션에서 처리하도록 수정했습니다!

f-lab-maverick reviewed (changes requested):
고생하셨습니다. 리뷰 확인해주시고, 수정하신 후 다시 요청해주세요~

codesejin committed:
refactor : Controller에서 @ResponseStatus삭제

codesejin committed:
refactor : Product Exception 추가 및 가격 타입 decimal변경

codesejin committed:
refactor : 메소드 네이밍 변경 - 쿠폰을 사용했다는 명시적인 네이밍

codesejin committed:
refactor : 주문 시 사용 가능한 쿠폰 조회 기능 리팩토링
쿼리에서 로직 삭제 후 애플리케이션에서 데이터 가공

codesejin committed:
refactor : 쿼리에서 로직 지우기

codesejin committed:
refactor : 람다 표현식에서 외부 변수 접근 문제 해결
기존 구현에서 람다 표현식이 외부 변수 accumulatedDiscountPrice를 수정하여 "Variable used in lambda expression should be final or effectively final" 오류가 발생했습니다. AtomicReference 래퍼를 사용하여 accumulatedDiscountPrice 변수를 안전하게 업데이트하도록 수정했습니다.

codesejin committed:
refactor : 소나클라우드 이슈 해결

codesejin reviewed (commented):
시큐리티 이슈를 아직 해결하지 못해서 parameter manipulation 공격에 대한 내용 제외하고 수정했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/domain/entity/helper/AvailableCouponInfo.java:
```diff
@@ -0,0 +1,108 @@
+        long discountPrice = vo.discountPrice() == null ? 0 : vo.discountPrice();
+        if (isSalePriceOverThanZero(vo.salePrice())) {
+            return calculateDiscountPrice(vo.discountType(), vo.salePrice(), discountRate, discountPrice);
+        } else {
+            return calculateDiscountPrice(vo.discountType(), vo.originalPrice(), discountRate, discountPrice);
+        }
```
else문 없이도 정상 동작 할 것으로 보이네요~ else, else if문은 읽을 때 개발자의 피로감을 늘릴 수 있기 때문에 안쓰는 방법을 연습하는게 좋습니다.
```suggestion
        } 
        return calculateDiscountPrice(vo.discountType(), vo.originalPrice(), discountRate, discountPrice);
        
```

codesejin replied:
넵 early return으로 변경했습니다!

f-lab-maverick commented on src/main/java/com/flab/offcoupon/service/coupon_use/OrderService.java:
```diff
@@ -0,0 +1,189 @@
+
+        // 2. 현재 시간이 쿠폰의 유효기간 범위내에 있는지 확인
+        List<CouponValidationPeriodVo> isBetweenValidatePeriodVo = couponRepository.getCouponValidationPeriod(request.getCouponId());
+        for (CouponValidationPeriodVo validationPeriod : isBetweenValidatePeriodVo) {
+            if (!isNowBetweenValidatePeriod(now, validationPeriod)) {
+                throw new CouponUsageInvalidPeriodException(COUPON_USAGE_INVALID_PERIOD
+                        .formatted(validationPeriod.couponId(), validationPeriod.validateStartDate(), validationPeriod.validateEndDate()));
+            }
+        }
```
이런 부분이 stream의 `anymatch`를 적용하면 훨씬 깔끔하게 코드가 변경됩니다~

codesejin replied:
넵 stream API사용하는걸로 변경했습니다!

f-lab-maverick reviewed (approved):
고생하셨습니다~ 코멘트 확인해서 수정 후 머지해주세요.

codesejin reviewed (commented):
피드백 반영해서 수정했습니다~

codesejin committed:
refactor : early return & stream API 적용
